requirements.txt

```diff
@@ -3,7 +3,10 @@ dipy
 # https://github.com/tensorflow/tensorflow/issues/44467
 h5py~=2.10.0
 Keras==2.3.1
-ivadomed
+# PyTorch's Linux distribution is very large due to its GPU support,
+# but we only need that for training models. Our users only need CPU.
+--extra-index-url https://download.pytorch.org/whl/cpu
+ivadomed~=2.9.6
```

Comment:
Previously, `ivadomed` was unpinned, so we would have gotten e.g. `2.10.0` when the `ivadomed` team decided to bump the minor version. But, with `~=`, we're locked to `2.9.X`. 

Was there any specific motivation for going with `~=` as opposed to `>=`?

Reply:
Because it's supposed to be [the safest](https://peps.python.org/pep-0440/#compatible-release)

> It matches any candidate version that is expected to be compatible with the specified version.

Is that actually true? It depends on how closely ivadomed follows semver.

But in any case it should balance between having to manually update pinnings every week and having surprise breakage in the future.

There's also the future to think about: if we just depend on `ivadomed` then someone trying to use SCT 5.4 in 3 years will get an ivadomed that's 3 years newer, with its APIs all shifted around probably.

Reply:
> There's also the future to think about: if we just depend on `ivadomed` then someone trying to use SCT 5.4 in 3 years will get an ivadomed that's 3 years newer, with its APIs all shifted around probably.

Just thinking out loud, there should be a strong reason to use an older version of SCT, as a matter of fact for any library unless it's LTS

Reply:
> There's also the future to think about: if we just depend on `ivadomed` then someone trying to use SCT 5.4 in 3 years will get an ivadomed that's 3 years newer, with its APIs all shifted around probably.

This isn't quite true, as SCT's stable releases use frozen dependencies. When 5.4 was created, for that specific version, `ivadomed` was pinned to:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/55fbd43cfee489486ea91f502b0117d7063732a0/requirements-freeze.txt#L34

To be clear: The SCT project maintains two sets of requirements, that are roughly equivalent to [`install_requires` vs. `requirements.txt`](https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/)

One set is very loosely pinned:

> It is not considered best practice to use install_requires to pin dependencies to specific versions, or to specify sub-dependencies (i.e. dependencies of your dependencies). This is overly-restrictive, and prevents the user from gaining the benefit of dependency upgrades.

The other is precisely pinned:

> Whereas install_requires requirements are minimal, requirements files often contain an exhaustive listing of pinned versions for the purpose of achieving [repeatable installations](https://pip.pypa.io/en/latest/user_guide/#repeatability) of a complete environment.

So, my thought process for pinning dependencies has always been to try to leave them unpinned on `master` (in order to regularly test upstream dependencies), while also freezing them in place for stable releases.

Reply:
In this case, since our default requirements file is being fed into `install_requires`, I would want to follow PyPA's recommendations:

> Additionally, it’s best practice to indicate any known lower or upper bounds. 
>
> For example, it may be known, that your project requires at least v1 of ‘A’, and v2 of ‘B’, so it would be like so:
> 
> ```
> install_requires=[
>    'A>=1',
>    'B>=2'
> ]
> ```

Since on master, from this point on, `ivadomed=2.9.6` is now a lower bound, I would want to follow their example and pin `ivadomed>=2.9.6`.

Reply:
> Just thinking out loud, there should be a strong reason to use an older version of SCT, as a matter of fact for any library unless it's LTS

We're thinking about scientific reproducibility. We want people in 20 years to be able to redo our studies. Or at least have a high chance of making them work.

> The SCT project maintains two sets of requirements, that are roughly equivalent to [`install_requires` vs. `requirements.txt`](https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/)

We do, but no one else does. Surely we're going against some convention with that choice What does npm do? What does golang do? `apt`? `pacman`?

> I would want to follow PyPA's recommendations:

I don't know if I trust PyPA's own recommendations. They can't seem to agree on a lot of things. For example, the [setuptools docs don't make any recommendations](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html). Why did they add `~=` and call it the "compatible" clause if no one should use it? Isn't the point of a major version upgrade supposed to be that there's breaking changes and that downstream should watch out?

Though I do see `>=` used in [jaraco/skeleton](https://github.com/jaraco/skeleton/blob/1a6b828304e7a8896b55d9ebf83f481ba7ebd568/setup.cfg#L29-L43), which is afaict the closest thing that that exists to documenting modern best practices.

Ah I don't know. I can't get my head around how to use pip properly. (and I know I'm not alone). Let's just do whatever works. I guess `>=` because jaraco says it's right.